File: Examples/ObjectiveC/AdMobNativeAdapterExample/AdMobNativeAdapterExample/ViewController.m
/*
 * Version for iOS © 2015–2021 YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://yandex.com/legal/mobileads_sdk_agreement/
 */

#import <GoogleMobileAds/GoogleMobileAds.h>
#import "ViewController.h"
#import "NativeAdView.h"

@interface ViewController () <GADNativeAdLoaderDelegate>

@property (nonatomic, strong) GADAdLoader *adLoader;
@property (nonatomic, weak) NativeAdView *adView;

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    [self configureAdLoader];
}

- (IBAction)loadAd:(UIButton *)sender
{
    [self configureAdView];
    [self.adLoader loadRequest:[GADRequest request]];
}

- (void)configureAdView
{
    NativeAdView *adView =
        [[[NSBundle mainBundle] loadNibNamed:@"NativeAdView" owner:nil options:nil] firstObject];
    adView.hidden = YES;
    [self addView:adView];
    [self.adView removeFromSuperview];
    self.adView = adView;
}

- (void)addView:(UIView *)view
{
    view.translatesAutoresizingMaskIntoConstraints = NO; 
    [self.view addSubview:view];
    NSDictionary *views = NSDictionaryOfVariableBindings(view);
    NSArray *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[view]-|"
                                                                             options:0
                                                                             metrics:nil
                                                                               views:views];
    NSArray *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[view]-|"
                                                                           options:0
                                                                           metrics:nil
                                                                             views:views];
    [self.view addConstraints:horizontalConstraints];
    [self.view addConstraints:verticalConstraints];
}

- (void)configureAdLoader
{
    // Replace ca-app-pub-XXXXXXXXXXXXXXXX/YYYYYYYYYY with Ad Unit ID generated at https://apps.admob.com".
    self.adLoader = [[GADAdLoader alloc] initWithAdUnitID:@"ca-app-pub-XXXXXXXXXXXXXXXX/YYYYYYYYYY"
                                       rootViewController:self
                                                  adTypes:@[ kGADAdLoaderAdTypeNative ]
                                                  options:nil];
    self.adLoader.delegate = self;
}

#pragma mark - GADNativeAdLoaderDelegate

- (void)adLoader:(GADAdLoader *)adLoader didFailToReceiveAdWithError:(NSError *)error
{
    NSLog(@"Ad loader did fail to receive ad with error: %@", error);
}

- (void)adLoader:(GADAdLoader *)adLoader didReceiveNativeAd:(GADNativeAd *)nativeAd
{
    NSLog(@"Ad loader did receive native ad");
    self.adView.nativeAd = nativeAd;
    [self.adView configureAssetViews];
    self.adView.hidden = NO;
}

@end

File: Examples/Swift/AdMobNativeAdapterExample/AdMobNativeAdapterExample/ViewController.swift
/*
 * Version for iOS © 2015–2021 YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://yandex.com/legal/mobileads_sdk_agreement/
 */

import GoogleMobileAds

class ViewController: UIViewController {
    private var adLoader: GADAdLoader!
    private var adView: NativeAdView?

    override func viewDidLoad() {
        super.viewDidLoad()

        createAdView()
        createLoader()
    }

    @IBAction func loadAd(_ sender: UIButton) {
        adLoader.load(GADRequest())
    }
    
    private func createAdView() {
        adView = Bundle.main.loadNibNamed("NativeAdView",
                                          owner: nil,
                                          options: nil)?.first as? NativeAdView
        if let adView = adView {
            adView.isHidden = true
            addView(adView)
        }
    }

    private func addView(_ adView: UIView) {
        adView.translatesAutoresizingMaskIntoConstraints = false
        self.view.addSubview(adView)
        var layoutGuide = self.view.layoutMarginsGuide
        if #available(iOS 11.0, *) {
            layoutGuide = self.view.safeAreaLayoutGuide
        }
        let constraints = [
            adView.leftAnchor.constraint(equalTo: layoutGuide.leftAnchor),
            adView.rightAnchor.constraint(equalTo: layoutGuide.rightAnchor),
            adView.bottomAnchor.constraint(equalTo: layoutGuide.bottomAnchor),
        ]
        NSLayoutConstraint.activate(constraints)
    }

    private func createLoader() {
        // Replace ca-app-pub-XXXXXXXXXXXXXXXX/YYYYYYYYYY with Ad Unit ID generated at https://apps.admob.com".
        adLoader = GADAdLoader(adUnitID: "ca-app-pub-XXXXXXXXXXXXXXXX/YYYYYYYYYY",
                               rootViewController: self,
                               adTypes: [.native],
                               options: nil)
        adLoader.delegate = self
    }
}

extension ViewController: GADNativeAdLoaderDelegate, GADNativeAdDelegate {
    func adLoader(_ adLoader: GADAdLoader, didReceive nativeAd: GADNativeAd) {
        guard let adView = adView else { return }

        nativeAd.delegate = self
        adView.nativeAd = nativeAd
        adView.configureAssetViews()
        adView.isHidden = false
    }

    func adLoader(_ adLoader: GADAdLoader, didFailToReceiveAdWithError error: Error) {
        print("Ad loader did fail to receive ad with error: \(error.localizedDescription)")
    }
}
